core/utils/utils.go

```diff
@@ -1112,3 +1114,47 @@ func MinKey[U any, T constraints.Ordered](elems []U, key func(U) T) T {
 
 	return min
 }
+
+// ErrorBuffer uses joinedErrors interface to join multiple errors into a single error.
+// This is useful to track the most recent N errors in a service and flush them as a single error.
+
+type ErrorBuffer struct {
+	// buffer is a slice of errors
+	buffer []error
+	// Cap is the maximum number of errors that the buffer can hold.
+	// Exceeding the cap results in discarding the oldest error
+	Cap int
+
+	sync.RWMutex
+}
+
+func (eb *ErrorBuffer) Flush() (err error) {
+	eb.RLock()
+	defer eb.RUnlock()
+	err = stderrs.Join(eb.buffer...)
+	eb.buffer = nil
+	return
+}
+
+func (eb *ErrorBuffer) Append(incoming error) {
+	eb.Lock()
+	defer eb.Unlock()
+	// if at capacity, drop first element
+	if len(eb.buffer) == eb.Cap && eb.Cap != 0 {
+		eb.buffer = append(eb.buffer[1:], incoming)
+		return
+	}
+	eb.buffer = append(eb.buffer, incoming)
```

Comment:
i think there is a nil pointer seg fault here if you call Flush then Append. on #1135 probably better to reset the buffer to empty rather than making nil

Reply:
built-in append does all the magic to reallocate underlying arrays if passed param is nil, only restriction is that it has to be a typed nil

https://goplay.tools/snippet/1XoqAUsmzri

Reply:
cool. thank you.